Add Cce H5 file type

knelli2 opened the pull request:
## Proposed changes

This new H5 file type will help streamline the usage of waveform output from the CCE executable within the [sxs](https://github.com/sxs-collaboration/sxs) and [scri](https://github.com/moble/scri) python packages.

These new Cce subfile is basically just a group that holds datasets for the bondi variables. The group and the datasets have attributes that the python packages will look for.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Factor out some H5 dat internals

knelli2 committed:
Add Cce H5 file type

wthrowe commented on src/IO/H5/Cce.cpp:
```diff
@@ -0,0 +1,262 @@
+}
+
+Cce::~Cce() {
+#ifdef __clang__
```
Why are we avoiding structured bindings with clang and using them with gcc?  If the point of this `#ifdef` is just to protect the pragmas, just put it around them and use the same normal c++ for both versions, otherwise comment what's going on.

knelli2 replied:
My compiler (clang 14) complained about trying to capture structured bindings in a lambda (in the ERROR macro I believe, which CHECK_H5 calls). GCC didn't care about this. The `#ifdef` was more for the pragmas. I can just avoid the structured bindings, add a comment explaining why, and only add the `#ifdef` for the pragmas.

nilsdeppe replied:
I think there is a C++17-ish bug about this that gets fixed in C++20 (or 23), so whether or not we can directly capture struct bindings in lambdas currently depends on the compiler :(

wthrowe replied:
Yes, just `#ifdef` the pragmas.

wthrowe commented on src/IO/H5/Cce.cpp:
```diff
@@ -0,0 +1,262 @@
+// The pragma here is used to suppress the warning that the compile time branch
+// of `ERROR` will always call `terminate()` because it throws an error.
+// Throwing that error is a code path that will never actually be entered at
+// runtime, so we suppress the warning here.
```
Why can't it happen?

knelli2 replied:
To be honest, I just copied this from the `Dat.cpp` file, so not really sure. @macedo22 added this when she made the ERROR macros usable at compile time.

macedo22 replied:
The warning being suppressed is for the branch taken in `ERROR` when we have a constant evaluated context (the code is being evaluated at compile time). This branch of `ERROR`, by construction, will never run at runtime because in this branch, we will get the error thrown at compile time.

macedo22 replied:
See the branch [here](https://github.com/sxs-collaboration/spectre/blob/a9ee6aa7c65ba703dbdc6dfb79436bb5378d4465/src/Utilities/ErrorHandling/Error.hpp#L68)

macedo22 replied:
At least, that is my understanding from when we made this change

wthrowe replied:
Ah, right.  That's fine then.

wthrowe commented on src/IO/H5/Cce.cpp:
```diff
@@ -0,0 +1,262 @@
+
+  // For a given ell, we nedd the first column index and the last column index.
+  // For a given ell, the total number of coefs is given by
+  // f(l) = 2 * (l + 1)^1 (first factor of 2 comes from having both real and
```
`^1` -> `^2`

wthrowe commented on tests/Unit/IO/H5/Test_Cce.cpp:
```diff
@@ -0,0 +1,579 @@
+#include "Utilities/MakeString.hpp"
+#include "Utilities/TMPL.hpp"
+
+#include "Parallel/Printf/Printf.hpp"
```
Remove.

wthrowe commented on tests/Unit/IO/H5/Test_Cce.cpp:
```diff
@@ -0,0 +1,579 @@
+  ss << std::regex_replace(build_info, std::regex{"\n"}, "\n# ");
+  const auto& header = cce_file.get_header();
+  CHECK("#\n# File created on "s ==
+        header.substr(0, header.find("File created on ") + 16));
```
I think this is just `header.starts_with("#\n# File created on ")`

wthrowe commented on tests/Unit/IO/H5/Test_Cce.cpp:
```diff
@@ -0,0 +1,579 @@
+  const auto& header = cce_file.get_header();
+  CHECK("#\n# File created on "s ==
+        header.substr(0, header.find("File created on ") + 16));
+  CHECK(ss.str() == header.substr(header.find("# SpECTRE")));
```
`header.ends_with(ss.str())`?

knelli2 commented:
Hmm the MacOS CI failed for the test I added. Not really sure how to debug without a Mac...

wthrowe reviewed (changes requested):
The test segfaulted on the macOS build.  I didn't catch anything that would cause that reading the code, but it's probably a real failure.

nilsvu commented:
I can look at the segfault at macos

knelli2 commented:
Fixup doesn't include anything for MacOS failure.

wthrowe commented:
Fixups look good.  Just need to figure out the Mac thing.

nilsvu commented:
Here's a fix for the macOS issue:

```diff
diff --git a/src/IO/H5/Helpers.cpp b/src/IO/H5/Helpers.cpp
index 6a6b5b6ca5..cf8001cbf2 100644
--- a/src/IO/H5/Helpers.cpp
+++ b/src/IO/H5/Helpers.cpp
@@ -473,8 +473,15 @@ void write_to_attribute(const hid_t location_id, const std::string& name,
   const hid_t att_id = H5Acreate2(location_id, name.c_str(), h5_type<Type>(),
                                   space_id, h5p_default(), h5p_default());
   CHECK_H5(att_id, "Failed to create attribute '" << name << "'");
-  CHECK_H5(H5Awrite(att_id, h5_type<Type>(), static_cast<const void*>(&value)),
-           "Failed to write value: " << value);
+  if constexpr (std::is_same_v<Type, std::string>) {
+    const auto cstr = value.c_str();
+    CHECK_H5(H5Awrite(att_id, h5_type<Type>(), &cstr),
+             "Failed to write value: " << value);
+  } else {
+    CHECK_H5(
+        H5Awrite(att_id, h5_type<Type>(), static_cast<const void*>(&value)),
+        "Failed to write value: " << value);
+  }
   CHECK_H5(H5Aclose(att_id), "Failed to close attribute '" << name << "'");
   CHECK_H5(H5Sclose(space_id), "Unable to close dataspace");
 }
```

Note that we don't currently store any other single string attributes in H5 files, maybe because of issues like this. Even the input file is stores as single-element vector, which seems wrong. Maybe with this fix the input file can be stored as a plain string?

nilsvu commented:
Can you also add Python bindings for the CCE file type? I expect we'll want to access those in Python quite a lot.

knelli2 committed:
Write H5 string attribute as C-style string

This was found to cause issues on MacOS when trying to write a
single string to an attribute.

knelli2 committed:
Bind H5 Cce file in Python

knelli2 commented:
1. Rebased
2. Added new commit for writing string attributes using C-style string (did not change current uses like in input_source. That can be done in a different PR)
3. Squashed fixup into "Add Cce" commit
4. Added new commit that binds the Cce h5 file in python

wthrowe commented on src/IO/H5/Helpers.cpp:
```diff
@@ -473,8 +473,15 @@ void write_to_attribute(const hid_t location_id, const std::string& name,
   CHECK_H5(att_id, "Failed to create attribute '" << name << "'");
-  CHECK_H5(H5Awrite(att_id, h5_type<Type>(), static_cast<const void*>(&value)),
-           "Failed to write value: " << value);
+  if constexpr (std::is_same_v<Type, std::string>) {
+    const auto cstr = value.c_str();
+    CHECK_H5(H5Awrite(att_id, h5_type<Type>(), &cstr),
+             "Failed to write value: " << value);
```
I'm very confused.  This seems to work, but every example I can find online passes this function a `char *`, while this passes a `char **`.  What's going on?

knelli2 replied:
I tried removing the `&` and got a seg fault

wthrowe replied:
Yes, this is the form that seems to work.  I'm just wondering if you know of any documentation that actually explains how to use the function or an explanation of why every example I can find is apparently wrong.  SpEC does the same thing as this, which makes me happier.

knelli2 replied:
No I don't know of any documentation for this. I just copied this from Nils Vu's patch he sent, and didn't question it.

nilsdeppe replied:
I can't find any explicit HDF5 documentation on this. A couple examples I found online actually don't write the attribute out because they use `H5S_NULL` instead of `H5S_SCALAR`. Below is an example code if you want to try digging into this. My best guess is that because the `H5S_SCALAR` is a "scalar string" that you must pass a pointer to the string, i.e. a pointer to the pointer of characters. Here's the `H5Awrite` documentation https://docs.hdfgroup.org/hdf5/develop/group___h5_a.html#title37 with the `void * buf` being what I interpret as the "pointer to scalar" which if the scalar is of type `char*` then I guess you'd want a `char**`. In the below example, if I switch from `H5S_SCALAR` to `H5S_NULL` the code runs but `h5dump ./Attributes.h5` shows the attribute isn't written. If I try `H5S_SIMPLE` everything explodes with cryptic HDF5 internals errors that I don't understand. So what I can tell you is that empirically, this is the right way to write a null-terminated variable-length string as an HDF5 attribute. It may be different for datasets, I don't remember. I do remember years ago when writing the helpers that getting strings working was a pain because online examples were essentially useless...

```cpp
 #include <hdf5.h>
 #include <string>

 #define H5FILE_NAME "Attributes.h5"
 #define ATTR_NAME "VarLenAttr" /* Name of the string attribute */

 #define DATASET_RANK  1   /* Rank and size of the dataset  */
 #define DATASET_SIZE  7

 int main() {
   hid_t file, dataset, attr;       /* File and dataset identifiers */
   hid_t dataset_space, attr_space; /* Dataset's and Attribute's dataspace id */
   hid_t attr_type;                 /* Attribute type */
   herr_t status;                   /* Return value */

   hsize_t fdim[] = {DATASET_SIZE};
   std::string data{"ASD"};

   file = H5Fcreate(H5FILE_NAME, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);

   dataset_space = H5Screate(H5S_SIMPLE);
   status = H5Sset_extent_simple(dataset_space, DATASET_RANK, fdim, NULL);

   dataset = H5Dcreate2(file, "my_dataset", H5T_NATIVE_INT, dataset_space,
                        H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

   attr_space = H5Screate(H5S_SCALAR);
   attr_type = H5Tcopy(H5T_C_S1);
   status = H5Tset_size(attr_type, H5T_VARIABLE);
   status = H5Tset_strpad(attr_type, H5T_STR_NULLTERM);
   status = H5Tset_cset(attr_type, H5T_CSET_ASCII);

   attr = H5Acreate2(dataset, ATTR_NAME, attr_type, attr_space, H5P_DEFAULT,
                     H5P_DEFAULT);

   const char* attr_string = data.c_str();
   status = H5Awrite(attr, attr_type, &attr_string);
   H5Aclose(attr);
   H5Tclose(attr_type);
   H5Sclose(attr_space);

   H5Fclose(file);
 }
```


Is this enough for you to be okay with this part of the diff or is there some explicit change request you still have on this part?

wthrowe replied:
Thanks for the code.  Yes, I don't think anything needs to change for this.  I just spent a lot of time confused about how the interface worked and hoped someone else knew better than me.

nilsdeppe replied:
You're welcome! Okay, sonuds good! I tried digging into the HDF5 internals but got stuck where they try to use structs with function pointer members as a way of mocking classes: https://github.com/HDFGroup/hdf5/blob/18321dee40b21e19c95bf0ef2f61e3ec71233fe9/src/H5VLconnector.h#L837 So I never got to the final "write the string to disk" stage since I have no idea where the "constructor" for this class is :(

wthrowe commented on src/IO/H5/Helpers.cpp:
```diff
@@ -473,8 +473,15 @@ void write_to_attribute(const hid_t location_id, const std::string& name,
-           "Failed to write value: " << value);
+  if constexpr (std::is_same_v<Type, std::string>) {
+    const auto cstr = value.c_str();
+    CHECK_H5(H5Awrite(att_id, h5_type<Type>(), &cstr),
```
From the `h5_type` docs: "\note For strings, the returned type must be released with H5Tclose"

knelli2 replied:
This seems like it isn't always done in the `Helpers.cpp`

wthrowe replied:
Yes, `read_value_attribute` also gets it wrong.  The other calls aren't strings.

nilsdeppe replied:
Will is correct, this is a latent bug in the other code as well. It only matters for strings since we set some properties on the string to make it nicer to deal with.

knelli2 replied:
Ok so shall I change it here, or do a follow-up?

wthrowe replied:
Fix it in this PR.

knelli2 replied:
Squashed it in to the "Write H5" commit since it was only a few lines
